File: GithubDashboard/app/src/main/java/com/zimpo/githubdashboard/Api/GithubUserApi.java
package com.zimpo.githubdashboard.Api;

import android.content.Context;
import android.os.AsyncTask;
import android.os.Build;
import android.widget.Toast;

import com.android.volley.AuthFailureError;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.zimpo.githubdashboard.Object.User;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import androidx.annotation.RequiresApi;

public class GithubUserApi extends AsyncTask {

    Context context;
    ProcessFinish delegate = null;
    ArrayList<User> users = null;

    public interface ProcessFinish
    {
        void processFinish(ArrayList<User> users);
    }

    public GithubUserApi(Context context, ProcessFinish delegate)
    {
        this.context = context;
        this.delegate = delegate;
        this.users = new ArrayList<>();
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    @Override
    protected Object doInBackground(Object[] params) {
        // HTTP POST
        String search = (String) params[0];
        search = search.replace(' ', '+');
        String url = "https://api.github.com/search/users?q=" + search;
        RequestQueue requestQueue = Volley.newRequestQueue(context);
        JSONObject jsonObject = new JSONObject();
        //jsonObject.put("key1", "value1");
        //jsonObject.put("key2", "value2");
        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.GET, url,
                jsonObject, new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {
                try {
                    //JSONObject count = (JSONObject) response.get("total_count");
                    int count = response.getInt("total_count");
                    if (count < 0)
                    {
                        Toast.makeText(context, "Aucun utiliseur n'a été trouvé", Toast.LENGTH_LONG).show();
                        return;
                    }

                    JSONArray items = response.getJSONArray("items");
                    for (int i = 0; i < items.length(); i++)
                    {
                        JSONObject user = items.getJSONObject(i);
                        //ID
                        int id = user.getInt("id");
                        //LOGIN
                        String login = user.getString("login");
                        //AVATAR_URL
                        String avatar_url = user.getString("avatar_url");
                        //URL
                        String url = user.getString("url");
                        //REPOS_URL
                        String repos_url = user.getString("repos_url");

                        users.add(new User(id, login, avatar_url, url, repos_url));
                    }

                    delegate.processFinish(users);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                Toast.makeText(context, error.toString(), Toast.LENGTH_LONG).show();
                return;
            }
        });
        /* {
            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                final Map<String, String> headers = new HashMap<>();
                headers.put("Authorization", "Bearer BQA1f5XXzb4GkMWxsjQVyqhFSolnVLaAgXgLGG-MT3_BARxhRDWO0M2a3_rwIuYh95bHiXdu3mkcINLTQwxotM8LhtuS10Ghg2HqOWYL0z39R77l4kZBhSqS0SxGh3FMialRBTinehiWjA");
                return headers;
            }
        };*/
        requestQueue.add(jsonObjectRequest);
        return true;
    }

    @Override
    protected void onPostExecute(Object o) {
        super.onPostExecute(o);
    }
}

File: GithubDashboard/app/src/main/java/com/zimpo/githubdashboard/Api/GithubUserApi.kt
package com.zimpo.githubdashboard.Api

import android.content.Context
import android.os.AsyncTask
import android.widget.Toast
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.VolleyError
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import com.zimpo.githubdashboard.Adapter.RepoAdapter
import com.zimpo.githubdashboard.Model.User
import java.util.ArrayList
import com.squareup.picasso.Downloader
import org.json.JSONException
import org.json.JSONObject


class GithubUserApi(private val context: Context, private val delegate: ProcessFinish) : AsyncTask<String, Void, Void>() {

    val users = ArrayList<User>()

    interface ProcessFinish {
        fun processFinish(users: ArrayList<User>)

        companion object {
            inline operator fun invoke(crossinline op: (users: ArrayList<User>) -> Unit) =
                    object : ProcessFinish {
                        override fun processFinish(users: ArrayList<User>) = op(users)
                    }
        }
    }

    override fun doInBackground(vararg searches: String): Void? {
        // HTTP POST
        var search = searches[0]
        search = search.replace(' ', '+')
        val url = "https://api.github.com/search/users?q=$search"
        val requestQueue = Volley.newRequestQueue(context)
        val jsonObject = JSONObject()
        val jsonObjectRequest = JsonObjectRequest(Request.Method.GET, url,
                jsonObject, Response.Listener { response ->
            try {
                //JSONObject count = (JSONObject) response.get("total_count");
                val count = response.getInt("total_count")
                if (count < 0) {
                    Toast.makeText(context, "Aucun utiliseur n'a été trouvé", Toast.LENGTH_LONG).show()
                    return@Listener
                }

                val items = response.getJSONArray("items")
                for (i in 0 until items.length()) {
                    val user = items.getJSONObject(i)
                    //ID
                    val id = user.getInt("id")
                    //LOGIN
                    val login = user.getString("login")
                    //AVATAR_URL
                    val avatar_url = user.getString("avatar_url")
                    //URL
                    val url = user.getString("url")
                    //REPOS_URL
                    val repos_url = user.getString("repos_url")

                    users.add(User(id, login, avatar_url, url, repos_url))
                }

                delegate.processFinish(users)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }, Response.ErrorListener { error ->
            Toast.makeText(context, error.toString(), Toast.LENGTH_LONG).show()
            return@ErrorListener
        })
        requestQueue.add(jsonObjectRequest)
        return null
    }
}
